src/peer_digest.cc

```diff
@@ -448,7 +426,7 @@ static int
 peerDigestFetchReply(void *data, char *buf, ssize_t size)
 {
     DigestFetchState *fetch = (DigestFetchState *)data;
-    PeerDigest *pd = fetch->pd;
+    const auto pd = fetch->pd.get();
     assert(pd && buf);
```

Comment:
What guarantees fetch->pd validity in DigestFetchState callbacks? AFAICT from a quick check, the answer is "nothing; in this particular callback, this code will assert in some cases". What am I missing?

Reply:
I changed the first argument's type to DigestFetchState* (42de02) so that this function (and a couple of similar ones) would not look as a callback. We could have added also a comment about that this function must not be called before peerDigestFetchedEnough() that checks for fetch->pd validity.

Reply:
> We could have added also a comment that this function must not be called before peerDigestFetchedEnough() that checks for fetch->pd validity.

I would not do that. I think that comment applies to all foo(fetch) functions except for peerDigestFetchedEnough() itself. The digest check inside the latter is like ERR_CLOSING and ENTRY_ABORTED -- something every digest fetching callback has to check first, before proceeding with regular actions and something that should not be re-checked during those regular actions.

Also, peerDigestFetchReply() itself calls peerDigestFetchedEnough(), so the comment will look odd unless we write an additional explanation.

Long-term, I think we should replace peerDigestFetchedEnough() calls with several (mostly inlined) pieces of context-specific code. That function is trying to check everything in every possible context, which needlessly complicates its API and checking logic. However, refactoring this may be too invasive for this PR.